pennylane/measurements/classical_shadow.py

```diff
@@ -314,6 +315,106 @@ def process(self, tape, device):
+            ]
+        )
+        obs = obs_list[recipes]
+        uni = diag_list[recipes]
+
+        # There's a significant speedup if we use the following iterative
+        # process to perform the randomized Pauli measurements:
+        #   1. Randomly generate Pauli observables for all snapshots for
+        #      a single qubit (e.g. the first qubit).
+        #   2. Compute the expectation of each Pauli observable on the first
+        #      qubit by tracing out all other qubits.
+        #   3. Sample the first qubit based on each Pauli expectation.
+        #   4. For all snapshots, determine the collapsed state of the remaining
+        #      qubits based on the sample result.
+        #   4. Repeat iteratively until no qubits are remaining.
+        #
+        # Observe that after the first iteration, the second qubit will become the
+        # "first" qubit in the process. The advantage to this approach as opposed to
+        # simulataneously computing the Pauli expectations for each qubit is that
+        # the partial traces are computed over iteratively smaller subsystems, leading
+        # to a significant speed-up.
+
+        # transpose the state so that the measured wires appear first
+        unmeasured_wires = [i for i in range(num_dev_qubits) if i not in mapped_wires]
+        transposed_state = np.transpose(state, axes=mapped_wires.tolist() + unmeasured_wires)
+
+        outcomes = np.zeros((shots, n_qubits))
+        stacked_state = np.stack([transposed_state for _ in range(shots)])
+
+        for active_qubit in range(n_qubits):
+            # trace out every qubit except the first
+            first_qubit_state = np.einsum(
+                f"{ABC[num_dev_qubits - active_qubit + 1]}{ABC[:num_dev_qubits - active_qubit]},"
+                f"{ABC[num_dev_qubits - active_qubit + 1]}{ABC[num_dev_qubits - active_qubit]}{ABC[1:num_dev_qubits - active_qubit]}"
+                f"->{ABC[num_dev_qubits - active_qubit + 1]}a{ABC[num_dev_qubits - active_qubit]}",
+                stacked_state,
+                np.conj(stacked_state),
+            )
+
+            # sample the observables on the first qubit
+            probs = (np.einsum("abc,acb->a", first_qubit_state, obs[:, active_qubit]) + 1) / 2
+            samples = bit_rng.random(size=probs.shape) > probs
+            outcomes[:, active_qubit] = samples
+
+            # collapse the state of the remaining qubits; the next qubit in line
+            # becomes the first qubit for the next iteration
+            rotated_state = np.einsum("ab...,acb->ac...", stacked_state, uni[:, active_qubit])
+            stacked_state = rotated_state[np.arange(shots), samples.astype(np.int8)]
+
+            # re-normalize the collapsed state
+            norms = np.sqrt(
+                np.sum(
+                    np.abs(stacked_state) ** 2,
+                    tuple(range(1, num_dev_qubits - active_qubit)),
+                    keepdims=True,
+                )
+            )
```

Comment:
This might be another candidate for breaking an extremely long line up into multiple lines with explanatory variables.

```suggestion
            sum_indices = tuple(range(1, num_remaining_qubits))
            state_squared = np.abs(stacked_state) ** 2,
            norms = np.sqrt(np.sum(state_squared, sum_indices, keepdims=True))
```

This might be another candidate to break up into multiple lines with explanatory variables. It can make code much easier to read if each line is only *doing* one (or two) things.